subcommand/create-federation-secret/command_test.go

```diff
@@ -0,0 +1,1118 @@
+	k8s := fake.NewSimpleClientset()
+
+	// Set up Consul server with TLS.
+	caFile, certFile, keyFile, cleanup := common.GenerateServerCerts(t)
+	defer cleanup()
+	a, err := testutil.NewTestServerConfigT(t, func(c *testutil.TestServerConfig) {
+		c.CAFile = caFile
+		c.CertFile = certFile
+		c.KeyFile = keyFile
+	})
+	require.NoError(t, err)
+	defer a.Stop()
+
+	// Create a mesh gateway instance after a delay.
+	meshGWIP := "192.168.0.1"
+	meshGWPort := 443
+	go func() {
+		time.Sleep(500 * time.Millisecond)
+		client, err := api.NewClient(&api.Config{
+			Address: a.HTTPAddr,
+		})
+		require.NoError(t, err)
+		err = client.Agent().ServiceRegister(&api.AgentServiceRegistration{
+			Name: "mesh-gateway",
+			TaggedAddresses: map[string]api.ServiceAddress{
+				"wan": {
+					Address: meshGWIP,
+					Port:    meshGWPort,
+				},
+			},
+		})
+		require.NoError(t, err)
+	}()
+
+	// Run the command.
+	ui := cli.NewMockUi()
+	cmd := Command{
+		UI:        ui,
+		k8sClient: k8s,
+	}
+	k8sNS := "default"
+	resourcePrefix := "prefix"
+	exitCode := cmd.Run([]string{
+		"-resource-prefix", resourcePrefix,
+		"-k8s-namespace", k8sNS,
+		"-mesh-gateway-service-name=mesh-gateway",
+		"-ca-file", caFile,
+		"-server-ca-cert-file", certFile,
+		"-server-ca-key-file", keyFile,
+		"-http-addr", fmt.Sprintf("https://%s", a.HTTPSAddr),
+	})
+	require.Equal(t, 0, exitCode, ui.ErrorWriter.String())
+
+	// Check the secret is as expected.
+	secret, err := k8s.CoreV1().Secrets(k8sNS).Get(resourcePrefix+"-federation", metav1.GetOptions{})
+	require.NoError(t, err)
+
+	// Test server config.
+	require.Contains(t, secret.Data, "serverConfigJSON")
+	expCfg := fmt.Sprintf(`{"primary_datacenter":"dc1","primary_gateways":["%s:%d"]}`, meshGWIP, meshGWPort)
+	require.Equal(t, expCfg, string(secret.Data["serverConfigJSON"]))
+}
+
+// Test when the mesh gateways don't have a tagged address of name "wan".
+func TestRun_MeshGatewayNoWANAddr(t *testing.T) {
+	t.Parallel()
+
+	// Set up Consul server with TLS.
+	caFile, certFile, keyFile, cleanup := common.GenerateServerCerts(t)
+	defer cleanup()
+	a, err := testutil.NewTestServerConfigT(t, func(c *testutil.TestServerConfig) {
+		c.CAFile = caFile
+		c.CertFile = certFile
+		c.KeyFile = keyFile
+	})
+	require.NoError(t, err)
+	defer a.Stop()
+	client, err := api.NewClient(&api.Config{
+		Address: a.HTTPAddr,
+	})
+	require.NoError(t, err)
+	err = client.Agent().ServiceRegister(&api.AgentServiceRegistration{
+		Name: "mesh-gateway",
+	})
+	require.NoError(t, err)
+
+	ui := cli.NewMockUi()
+	k8s := fake.NewSimpleClientset()
+	cmd := Command{
+		UI:        ui,
+		k8sClient: k8s,
+	}
+	exitCode := cmd.Run([]string{
+		"-resource-prefix=prefix",
+		"-k8s-namespace=default",
+		"-mesh-gateway-service-name=mesh-gateway",
+		"-ca-file", caFile,
+		"-server-ca-cert-file", caFile,
+		"-server-ca-key-file", keyFile,
+		"-http-addr", fmt.Sprintf("https://%s", a.HTTPSAddr),
+	})
+	require.Equal(t, 1, exitCode, ui.ErrorWriter.String())
```

Comment:
Do we want to check error message here?

Reply:
It's the logger that logs it so it doesn't show up here.